server/datastore/mysql/certificate_templates.go

```diff
@@ -0,0 +1,179 @@
+	}
+	if len(templates) > perPage {
+		templates = templates[:perPage]
+		paginationMetaData.HasNextResults = true
+	}
+
+	return templates, paginationMetaData, nil
+}
+
+func (ds *Datastore) CreateCertificateTemplate(ctx context.Context, certificateTemplate *fleet.CertificateTemplate) (*fleet.CertificateTemplateResponseFull, error) {
+	result, err := ds.writer(ctx).ExecContext(ctx, `
+		INSERT INTO certificate_templates (
+			name,
+			team_id,
+			certificate_authority_id,
+			subject_name
+		) VALUES (?, ?, ?, ?)
+	`, certificateTemplate.Name, certificateTemplate.TeamID, certificateTemplate.CertificateAuthorityID, certificateTemplate.SubjectName)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "inserting certificate_template")
+	}
+
+	id, err := result.LastInsertId()
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "getting last insert id for certificate_template")
+	}
+
+	return &fleet.CertificateTemplateResponseFull{
+		CertificateTemplateResponseSummary: fleet.CertificateTemplateResponseSummary{
+			ID:                     uint(id), //nolint:gosec
+			Name:                   certificateTemplate.Name,
+			CertificateAuthorityId: certificateTemplate.CertificateAuthorityID,
+		},
+		SubjectName: certificateTemplate.SubjectName,
+		TeamID:      certificateTemplate.TeamID,
+	}, nil
+}
+
+func (ds *Datastore) DeleteCertificateTemplate(ctx context.Context, id uint) error {
+	result, err := ds.writer(ctx).ExecContext(ctx, `
+		DELETE FROM certificate_templates
+		WHERE id = ?
+	`, id)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "deleting certificate_template")
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "getting rows affected for certificate_template")
+	}
+	if rowsAffected == 0 {
+		return notFound("CertificateTemplate").WithID(id)
+	}
+
+	return nil
+}
+
+func (ds *Datastore) BatchUpsertCertificateTemplates(ctx context.Context, certificateTemplates []*fleet.CertificateTemplate) error {
+	if len(certificateTemplates) == 0 {
+		return nil
+	}
+
+	const argsCountInsertCertificate = 4
+
+	const sqlInsertCertificate = `
+		INSERT INTO certificate_templates (
+			name,
+			team_id,
+			certificate_authority_id,
+			subject_name
+		) VALUES %s
+		ON DUPLICATE KEY UPDATE
+			name = VALUES(name),
+			team_id = VALUES(team_id),
+			certificate_authority_id = VALUES(certificate_authority_id),
+			subject_name = VALUES(subject_name)
+	`
+
+	var placeholders strings.Builder
+	args := make([]interface{}, 0, len(certificateTemplates)*argsCountInsertCertificate)
+
+	for _, cert := range certificateTemplates {
+		args = append(args, cert.Name, cert.TeamID, cert.CertificateAuthorityID, cert.SubjectName)
+		placeholders.WriteString("(?,?,?,?),")
+	}
+
+	stmt := fmt.Sprintf(sqlInsertCertificate, strings.TrimSuffix(placeholders.String(), ","))
+
+	if _, err := ds.writer(ctx).ExecContext(ctx, stmt, args...); err != nil {
+		return ctxerr.Wrap(ctx, err, "upserting certificate_templates")
+	}
```

Comment:
nit: If we add `db` tags to the `CertificateTemplate` struct type, we can use `NamedExecContext` here:
```suggestion
	const sqlInsertCertificate = `
        INSERT INTO certificate_templates (
            name,
            team_id,
            certificate_authority_id,
            subject_name
        ) VALUES (:name, :team_id, :certificate_authority_id, :subject_name)
        ON DUPLICATE KEY UPDATE
            name = VALUES(name),
            team_id = VALUES(team_id),
            certificate_authority_id = VALUES(certificate_authority_id),
            subject_name = VALUES(subject_name)
    `

	if _, err := ds.writer(ctx).NamedExecContext(ctx, sqlInsertCertificate, certificateTemplates); err != nil {
		return ctxerr.Wrap(ctx, err, "upserting certificate_templates")
	}
```